packages/flutter/lib/src/services/text_editing_delta.dart

```diff
@@ -126,14 +137,23 @@ abstract class TextEditingDelta {
     );
 
     if (isNonTextUpdate) {
+      assert(_debugTextRangeIsValid(newSelection, oldText), 'TextEditingDelta.fromJSON failed, the selection range: $newSelection is not within the bounds of text: $oldText of length: ${oldText.length}');
```

Comment:
Nit: I think you could just start the error message at "the selection range..." if you want because the first part is redundant.  The stacktrace will point to `TextEditingDelta.fromJSON` and "failed" is also obvious.